_posts/2022-10-27-kani-bolero.md

```diff
@@ -0,0 +1,317 @@
+---
+layout: post
+title: "From Fuzzing to Proof: Using Kani with the Bolero Property-Testing Framework"
+---
+
+Today we're going to talk about how you can use fuzzing and verification in a unified framework, which is enabled by the integration of the [Kani Rust Verifier](https://model-checking.github.io/kani/) in [Bolero](https://camshaft.github.io/bolero/).
+Bolero is a property-testing framework that makes it easy for users to test a piece of Rust code with multiple fuzzing engines, and Kani is a verification tool that allows users to prove certain properties about their Rust code.
+By integrating Kani in Bolero, we enable users to apply multiple verification methods, namely fuzzing and proofs, to the same harness.
+
+## Why Bolero?
+
+Suppose we want to test a function `update_account_balance` that takes the current balance and the amount we want to deposit or withdraw, and returns the new balance:
+
+```rust
+fn update_account_balance(current_balance: i32, amount: i32) -> i32 {
+    // compute the new balance ...
+}
+```
+
+A [property-based test](https://medium.com/criteo-engineering/introduction-to-property-based-testing-f5236229d237) harness would look as follows:
+
+```rust
+#[test]
+fn test_update_account_balance() {
+    let current_balance = /* some value */;
+    let amount = /* some value */;
+    let new_balance = update_account_balance(current_balance, amount);
+    assert!(balance_update_is_correct(current_balance, amount, new_balance));
+}
+```
+
+The question then is: how do we generate values for `current_balance` and `amount` for this property-based test harness?
+The answer depends on the testing technique we want to use.
+For example, if we were to use *random testing*, we would inject random values for `current_balance` and `amount`.
+If we do so using the [rand](https://rust-random.github.io/rand/rand/index.html) crate, the harness would be as follows:
+
+```rust
+fn test_update_account_balance_random() {
+    let current_balance = rand::random();
+    let amount = rand::random();
+    let result = update_account_balance(current_balance, amount);
+    assert!(balance_update_is_correct(current_balance, amount, result));
+}
+```
+
+We can then either repeatedly call the program to test it with different values, or call this function inside a loop.
+
+This form of random testing typically has limited effectiveness.
+We can improve the effectiveness of our randomized testing with grey-box fuzzers, which use metrics such as [code coverage](https://en.wikipedia.org/wiki/Code_coverage) to trigger unique behavior.
+There are a number of fuzzing engines for Rust, such as [libfuzzer](https://github.com/rust-fuzz/libfuzzer) and [AFL](https://github.com/rust-fuzz/afl.rs), that use different heuristics/metrics for exploring the input space.
+
+To use libfuzzer, we would write a harness that looks as follows:
```

Comment:
I'd use the name as written [here](https://llvm.org/docs/LibFuzzer.html) (libFuzzer)

Reply:
Done